src/coreclr/jit/codegenwasm.cpp

```diff
@@ -59,9 +59,110 @@ void CodeGen::genFuncletEpilog()
+void CodeGen::genCodeForTreeNode(GenTree* treeNode)
 {
-    NYI_WASM("genCodeForTreeNode");
+#ifdef DEBUG
+    lastConsumedNode = nullptr;
+    if (compiler->verbose)
+    {
+        compiler->gtDispLIRNode(treeNode, "Generating: ");
+    }
+#endif // DEBUG
+
+    assert(!treeNode->IsReuseRegVal()); // TODO-WASM-CQ: enable.
+
+    // Contained nodes are part of the parent for codegen purposes.
+    if (treeNode->isContained())
+    {
+        return;
+    }
+
+    switch (treeNode->OperGet())
+    {
+        case GT_ADD:
+            genCodeForBinary(treeNode->AsOp());
+            break;
+
+        case GT_LCL_VAR:
+            genCodeForLclVar(treeNode->AsLclVar());
+            break;
+
+        case GT_RETURN:
+            genReturn(treeNode);
+            break;
+
+        case GT_IL_OFFSET:
+            // Do nothing; this node is a marker for debug info.
+            break;
+
+        default:
+#ifdef DEBUG
+            NYIRAW(GenTree::OpName(treeNode->OperGet()));
+#else
+            NYI_WASM("Opcode not implemented");
+#endif
+            break;
+    }
+}
+
+//------------------------------------------------------------------------
+// genCodeForBinary: Generate code for a binary arithmetic operator
+//
+// Arguments:
+//    treeNode - The binary operation for which we are generating code.
+//
+void CodeGen::genCodeForBinary(GenTreeOp* treeNode)
+{
+    genConsumeOperands(treeNode);
+
+    instruction ins;
+    switch (treeNode->OperGet())
+    {
+        case GT_ADD:
+            if (!treeNode->TypeIs(TYP_INT))
+            {
+                NYI_WASM("genCodeForBinary: non-INT GT_ADD");
+            }
+            ins = INS_i32_add;
+            break;
+        default:
+            ins = INS_none;
+            NYI_WASM("genCodeForBinary");
+            break;
+    }
+
+    GetEmitter()->emitIns(ins);
+    genProduceReg(treeNode);
+}
+
+//------------------------------------------------------------------------
+// genCodeForLclVar: Produce code for a GT_LCL_VAR node.
+//
+// Arguments:
+//    tree - the GT_LCL_VAR node
+//
+void CodeGen::genCodeForLclVar(GenTreeLclVar* tree)
+{
+    assert(tree->OperIs(GT_LCL_VAR) && !tree->IsMultiReg());
+    LclVarDsc* varDsc = compiler->lvaGetDesc(tree);
+
+    // Unlike other targets, we can't "reload at the point of use", since that would require inserting instructions
+    // into the middle of an already-emitted instruction group. Instead, we order the nodes in a way that obeys the
+    // value stack constraints of WASM precisely. However, the liveness tracking is done in the same way as for other
+    // targets, hence "genProduceReg" is only called for non-candidates.
+    if (!varDsc->lvIsRegCandidate())
+    {
+        var_types type = varDsc->GetRegisterType(tree);
+        // TODO-WASM: actually local.get the frame base local here.
+        GetEmitter()->emitIns_S(ins_Load(type), emitTypeSize(tree), tree->GetLclNum(), 0);
+        genProduceReg(tree);
```

Comment:
I would have expected to see an emit with a memarg here, am I misunderstanding what this does?

Reply:
`emitIns_S` does emit an instruction with the local's offset (relative to the eventual FP) in memarg.

'`S`' (for `Stack`) operands in the emitter usually relate to address modes of the form `[FP/SP+<local offset>]`. Furthermore, they usually have dedicated `instrDesc`s, but here I chose to not do that since it's only going to affect disassembly display. We can revisit that if needed in the future.